core/module_graph.cc

```diff
@@ -0,0 +1,254 @@
+      }
+    }
+    node->parent_tasks_.push_back(parent);
+    return true;
+  }
+
+  auto node_it = module_graph_.find(node_name);
+  if (node_it == module_graph_.end()) {
+    return false;
+  }
+
+  for (auto &child_name : node_it->second.children()) {
+    if (visited->count(child_name) > 0) {
+      continue;
+    }
+    if (!FindNextTask(child_name, parent_name, visited)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Updates the parents of modules with tasks by traversing `module_graph_` and
+// ignoring all modules that are not tasks.
+bool ModuleGraph::UpdateTaskGraph() {
+  for (auto const &task : tasks_) {
+    std::unordered_set<std::string> visited;
+    if (!FindNextTask(task, task, &visited)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+const std::map<std::string, Module *> &ModuleGraph::GetAllModules() {
+  return all_modules_;
+}
+
+bool ModuleGraph::ExistModuleClass(const ModuleBuilder *builder) {
+  for (auto const &e : all_modules_) {
+    if (e.second->module_builder() == builder) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool ModuleGraph::AddEdge(const std::string &from, const std::string &to) {
+  auto from_it = module_graph_.find(from);
+  if (from_it == module_graph_.end() || module_graph_.count(to) == 0) {
+    return false;
+  }
+  from_it->second.AddChild(to);
+  return UpdateTaskGraph();
+}
+
+bool ModuleGraph::RemoveEdge(const std::string &from, const std::string &to) {
+  auto from_node = module_graph_.find(from);
+  if (from_node == module_graph_.end() || module_graph_.count(to) == 0) {
+    return false;
+  }
+
+  from_node->second.RemoveChild(to);
+
+  // We need to regenerate the task graph.
+  for (auto const &task : tasks_) {
+    auto it = all_modules_.find(task);
+    if (it != all_modules_.end()) {
+      it->second->parent_tasks_.clear();
+    }
+  }
+  return UpdateTaskGraph();
+}
+
+bool ModuleGraph::AddModule(Module *m) {
+  if (m->is_task_) {
+    if (!tasks_.insert(m->name()).second) {
+      return false;
+    }
+  }
+
+  bool module_added = all_modules_.insert({m->name(), m}).second;
+  if (!module_added) {
+    return false;
+  }
+
+  return module_graph_
+      .emplace(std::piecewise_construct, std::forward_as_tuple(m->name()),
+               std::forward_as_tuple(m))
+      .second;
+}
+
+int ModuleGraph::DestroyModule(Module *m, bool erase) {
+  int ret;
+  m->DeInit();
+
+  // disconnect from upstream modules.
+  for (size_t i = 0; i < m->igates_.size(); i++) {
+    ret = m->DisconnectModulesUpstream(i);
+    if (ret) {
+      return ret;
+    }
+  }
+
+  // disconnect downstream modules
+  for (size_t i = 0; i < m->ogates_.size(); i++) {
+    ret = m->DisconnectModules(i);
+    if (ret) {
+      return ret;
+    }
+  }
+
+  m->DestroyAllTasks();
+  m->DeregisterAllAttributes();
+
+  if (erase) {
+    all_modules_.erase(m->name());
+  }
+
+  module_graph_.erase(m->name());
+  if (m->is_task_) {
+    tasks_.erase(m->name());
+  }
+
+  delete m;
```

Comment:
`m` appears to only get deleted here; what happens if one of the other execution paths is taken?  If we need to delete in those cases too, maybe we can put m into a `std::unique_ptr` at the top?

Reply:
Yes, definitely it would be better if we make the ownership of modules (as well as TCs, tasks) much clearer.

Reply:
I am creating a new PR to address the issues handling module creation and deletion correctly. Not only with `DestroyModule()`, but also between `CreateModule()` and `AddModule()` has the same problem.